apps/src/javalab/redux/javalabRedux.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * Redux store for generic Java Lab state.
+ */
+
+import {PayloadAction, createSlice} from '@reduxjs/toolkit';
+
+interface JavalabState {
+  isRunning: boolean;
+  isTesting: boolean;
+  backpackEnabled: boolean;
+  isStartMode: boolean;
+  levelName: string | undefined;
+  isReadOnlyWorkspace: boolean;
+  hasOpenCodeReview: boolean;
+  isCommitSaveInProgress: boolean;
+  hasCommitSaveError: boolean;
+  validationPassed: boolean;
+  hasRunOrTestedCode: boolean;
+  isJavabuilderConnecting: boolean;
+}
+
+const initialState: JavalabState = {
+  isRunning: false,
+  isTesting: false,
+  backpackEnabled: false,
+  isStartMode: false,
+  levelName: undefined,
+  isReadOnlyWorkspace: false,
+  hasOpenCodeReview: false,
+  isCommitSaveInProgress: false,
+  hasCommitSaveError: false,
+  validationPassed: false,
+  hasRunOrTestedCode: false,
+  isJavabuilderConnecting: false
+};
+
+const javalabSlice = createSlice({
+  name: 'javalab',
+  initialState,
+  reducers: {
+    setIsRunning(state, action: PayloadAction<boolean>) {
+      state.isRunning = action.payload;
+    },
+    setIsTesting(state, action: PayloadAction<boolean>) {
+      state.isTesting = action.payload;
+    },
+    setBackpackEnabled(state, action: PayloadAction<boolean>) {
+      state.backpackEnabled = action.payload;
+    },
+    /**
+     * We should move isStartMode and levelName into a separate level redux file,
+     * or convert this design to one more closely matching redux/applab. When we
+     * do, we can remove the special treatment of Javalab in ImagePicker.jsx that
+     * enables the Asset Manager to run.
+     */
+    setIsStartMode(state, action: PayloadAction<boolean>) {
+      state.isStartMode = action.payload;
+    },
+    setLevelName(state, action: PayloadAction<string>) {
+      state.levelName = action.payload;
+    },
+    setIsJavabuilderConnecting(state, action: PayloadAction<boolean>) {
+      state.isJavabuilderConnecting = action.payload;
+    },
+    setIsReadOnlyWorkspace(state, action: PayloadAction<boolean>) {
+      state.isReadOnlyWorkspace = action.payload;
+    },
+    setHasOpenCodeReview(state, action: PayloadAction<boolean>) {
+      state.hasOpenCodeReview = action.payload;
+    },
+    setCommitSaveStatus: {
+      reducer(
+        state,
+        action: PayloadAction<{
+          isCommitSaveInProgress: boolean;
+          hasCommitSaveError: boolean;
+        }>
+      ) {
+        state.isCommitSaveInProgress = action.payload.isCommitSaveInProgress;
+        state.hasCommitSaveError = action.payload.hasCommitSaveError;
+      },
+      prepare(isCommitSaveInProgress: boolean, hasCommitSaveError: boolean) {
```

Comment:
So interestingly - it looks like in the associated view (CommitDialog.jsx), we do call this action with an object that matches this payload structure, but only convert the object into separate parameters in the [connect function](https://github.com/code-dot-org/code-dot-org/blob/87af4918db054aa6cf26d8cb7d7a7b84e33ef069/apps/src/javalab/CommitDialog.jsx#L338). So, if we do want to avoid using `prepare`, we could also just update the connect function to pass the object instead of separate args. This way works too though - just noticed this when I was looking at the file.

Reply:
oh weird. I'll simplify this then.